### PR TITLE
CRC support - Initial Implementation

### PR DESCRIPTION
**Pull request by robstryker**
Signed-off-by: Rob Stryker <rob@oxbeef.net>

# Pull Request Checklist
## General
* Is this a blocking issue or new feature? If yes, QE needs to +1 this PR

## Code
* Are method-/class-/variable-names meaningful?
* Are methods concise, not too long?
* Are catch blocks catching precise Exceptions only (no catch all)?

## Testing
* Are there unit-tests?
* Are there integration tests (or at least a jira to tackle these)?
* Is the non-happy path working, too?
* Are other parts that use the same component still working fine?

## Function
* Does it work?

### CODE REVIEW

File: plugins/org.jboss.tools.openshift.cdk.server/plugin.xml [hunk 1]
```diff
@@ -70,6 +70,22 @@
             stopTimeout="900000"
             supportsRemoteHosts="false"/>
 
+    <serverType
+            behaviourClass="org.jboss.tools.openshift.internal.cdk.server.core.adapter.CRCServerBehaviour"
```
Comment (jeffmaury) [inline plugins/org.jboss.tools.openshift.cdk.server/plugin.xml]: As we will one day or another remove CDK support, would be good to have the CRC support have its own namespace ie org.jboss.tools.openshift.internal.crc

**Comment by odockal** [comment]
What is nice is that server started after a little tackling. I am putting together a list of issues I run into.

**Comment by odockal** [comment]
I have tried the adapter and run into following issues:
1. Have prepared crc binary but do not have it set, yet
1. Create new CRC adapter
2. Set proper CRC binary and pull secret file
3. Start server
4. ASSERT: One is asked to setup it up first -> confirm
5. ASSERT: Log is being written to terminal output (you are asked to pass user credentials to configure network manager - I personally do not like this one, hope this could be configured before via cli)
6. ASSERT: Setup is done, no error in error log
Actual: I got error, "Server cannot be started unti CRC is has been initialized", it just pops up in background when confirming dialog for setting up the crc
7. ASSERT: Server is starting, etc..
Actual: Nothing happens after setup is done. 

Workaround: Start server again
ASSERT: Server is started

**Comment by odockal** [comment]
At the end of server adapter start up, when I got Openshift SSL cert acceptance dialog, just after confirming the dialog two same errors pops up in error log:
```
null
org.eclipse.core.jobs
Error
Tue Sep 17 16:05:38 CEST 2019
An internal error occurred during: "Configuring CRC Openshift Connection...".

com.openshift.restclient.OpenShiftException: Unable to execute request to https://api.crc.testing:6443/apis/user.openshift.io/v1/users/~
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:307)
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:275)
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:243)
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:226)
	at com.openshift.internal.restclient.DefaultClient.get(DefaultClient.java:414)
	at com.openshift.internal.restclient.authorization.AuthorizationContext.isAuthorized(AuthorizationContext.java:63)
	at org.jboss.tools.openshift.core.connection.Connection.authorize(Connection.java:237)
	at org.jboss.tools.openshift.core.connection.Connection.connect(Connection.java:226)
	at org.jboss.tools.openshift.internal.cdk.server.core.listeners.ConfigureCRCFrameworksListener.configureOpenshift(ConfigureCRCFrameworksListener.java:105)
	at org.jboss.tools.openshift.internal.cdk.server.core.listeners.ConfigureCRCFrameworksListener.configureFrameworks(ConfigureCRCFrameworksListener.java:69)
	at org.jboss.tools.openshift.internal.cdk.server.core.listeners.ConfigureCRCFrameworksListener$1.run(ConfigureCRCFrameworksListener.java:59)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
	at java.base/sun.security.ssl.SSLSocketImpl.handleEOF(SSLSocketImpl.java:1322)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1160)
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1063)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:402)
	at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.java:336)
	at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.java:300)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:185)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.java:224)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.java:108)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.java:88)
	at okhttp3.internal.connection.Transmitter.newExchange(Transmitter.java:169)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:41)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at com.openshift.internal.restclient.okhttp.ResponseCodeInterceptor.intercept(ResponseCodeInterceptor.java:55)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:221)
	at okhttp3.RealCall.execute(RealCall.java:81)
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:301)
	... 11 more
Caused by: java.io.EOFException: SSL peer shut down incorrectly
	at java.base/sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:167)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:108)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1152)
	... 37 more
```

**Comment by odockal** [comment]
@robstryker Can you take a look at this crc server wizard validation issues?
CRC Server wizard validation errors:
1. Open New Server wizard, choose CRC -> Next
2. There is an error marker when wizard page is opened, even though I think it should be present after user action - typing a path into a label. This is exactly how we have it solved in CDK wizard.

1. Open New Server wizard, choose CRC -> Next
2. Type some existing directory (ie. "/") into crc binary label
ASSERT: "Please select a valid image for crc" or similar
Actual: error marker pops up by CRC pull secret label and at page title area saying: "Please select a valid Image Pull Secret File"

1. Open New Server wizard, choose CRC -> Next
2. Insert crc binary path into crc binary label
ASSERT: Error marker pops up at secret file label and at page title area saying: "Please select a valid pull secret file" or similar
Actual: Error says: "Please ... valid Image Pull Secret File" - I would change the error message
3. Insert existing directory into secret file label 
ASSERT: Please select valid pull secret file
Actual: as above
4. Insert not existing path into secret file label
ASSERT: File does not exist
Actual: Please select...
5. Insert some file (pdf)
ASSERT: Please select valid pull secret file
Actual: we are ok - which is wrong

1. Open New Server wizard, choose CRC -> Next
2. Insert pull secret file into CRC pull secret file
ASSERT: Error marker pops up by CRC binary label and at page title saying: "Please select a valid Image Pull Secret File" OR Select file does not exist -> as label is empty now, it is ok
3. Insert existing path with directory
ASSERT: Please select valid image or similar
Actual: We are ok - which is wrong
4. Insert not existing path
ASSERT: Select file does not exist
5. Insert existing file (pdf)
ASSERT: Please select valid image...
Actual: we are ok- which is wrong

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
+package org.jboss.tools.openshift.internal.crc.server.ui.view;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.debug.core.ILaunch;
+import org.eclipse.debug.core.Launch;
+import org.eclipse.jdt.internal.launching.sourcelookup.advanced.AdvancedSourceLookupDirector;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java]: unnecessary import

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	public SetupCRCJob(IServer server, Shell shell, boolean useTerminal) {
+		this(server, shell, useTerminal, false);
+	}
+
+	public SetupCRCJob(IServer server, Shell shell, boolean useTerminal, boolean wait) {
+		super(server, shell, "Setup CRC", wait);
+		this.useTerminal = useTerminal;
+	}
+
+	protected String getContainerHome() {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java]: missing ```@Override```

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		return new File(System.getProperty("user.home"), ".crc").getAbsolutePath();
+	}
+	
+	protected boolean isValid() {
+		return (CRC100Server) server.loadAdapter(CRC100Server.class, new NullProgressMonitor()) != null;
+	}
+
+	protected String getBinaryLocation() {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java]: missing ```@Override```

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		return  BinaryUtility.CRC_BINARY.getLocation(server);
+	}
+	protected String getLaunchArgs() {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java]: missing ```@Override```

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		return "setup";
+	}
+	protected ILaunch launchSetup(IServer server) {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java]: missing ```@Override```

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.debug.core.ILaunch;
+import org.eclipse.debug.core.Launch;
+import org.eclipse.jdt.internal.launching.sourcelookup.advanced.AdvancedSourceLookupDirector;
+import org.eclipse.swt.widgets.Shell;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java]: missing class header

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
+package org.jboss.tools.openshift.internal.crc.server.ui.view;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/view/SetupCRCJob.java]: missing class header

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/CRC100ServerWizardFragment.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
+package org.jboss.tools.openshift.internal.crc.server.ui;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/CRC100ServerWizardFragment.java]: missing class header

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/cdk/server/core/adapter/controllers/ProcessUtility.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
+package org.jboss.tools.openshift.internal.cdk.server.core.adapter.controllers;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/cdk/server/core/adapter/controllers/ProcessUtility.java]: missing class header

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/cdk/server/core/adapter/controllers/StreamGobbler.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
+package org.jboss.tools.openshift.internal.cdk.server.core.adapter.controllers;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/cdk/server/core/adapter/controllers/StreamGobbler.java]: missing class header

**Comment by odockal** [comment]
Server Wizard validation: There is substantial improvement, thanks @robstryker. Nevertheless, I still find some issues:
Passing any executable file (sh script) into "CRC Binary:" labeled text passes the validation, which I think it should not. Should show error saying something like "Please select valid CRC binary/Image" whatever is more precise. 

Server Editor validation is not complete, I can play around with non-existing/directory/non-executable/executable-not-an-crc paths and I still can save the editor without any error/warning dialog  appearing.

**Comment by odockal** [comment]
@robstryker @adietish @jeffmaury  I would ilke to ask why are we using kubeadmin openshift connection be default... I am thinking that we might be using the other one - developer instead, as it was also used in developers.redhat.com blog about CRC. (https://developers.redhat.com/blog/2019/09/05/red-hat-openshift-4-on-your-laptop-introducing-red-hat-codeready-containers/) WDYT?

**Comment by odockal** [comment]
@adietish @robstryker I am also encountering probably lack of resources symptoms that are showing as one want to refresh OS connection and got:
```
null
org.jboss.tools.openshift.ui
Error
Thu Sep 19 14:42:10 CEST 2019
Failed to refresh element

com.openshift.restclient.OpenShiftException: Unable to execute request to https://api.crc.testing:6443/apis/image.openshift.io/v1/namespaces/openshift-monitoring/imagestreamtags
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:310)
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:275)
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:264)
	at com.openshift.internal.restclient.DefaultClient.list(DefaultClient.java:171)
	at com.openshift.internal.restclient.DefaultClient.list(DefaultClient.java:160)
	at com.openshift.internal.restclient.DefaultClient.list(DefaultClient.java:151)
	at org.jboss.tools.openshift.core.connection.Connection.getResources(Connection.java:475)
	at org.jboss.tools.openshift.internal.ui.models.ConnectionWrapper.refresh(ConnectionWrapper.java:289)
	at org.jboss.tools.openshift.internal.ui.models.ProjectWrapper.refresh(ProjectWrapper.java:109)
	at org.jboss.tools.openshift.internal.ui.models.ConnectionWrapper.refresh(ConnectionWrapper.java:273)
	at org.jboss.tools.openshift.internal.ui.handler.RefreshResourceHandler$1.doRun(RefreshResourceHandler.java:59)
	at org.jboss.tools.openshift.internal.common.core.job.AbstractDelegatingMonitorJob.run(AbstractDelegatingMonitorJob.java:37)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.net.ConnectException: Failed to connect to api.crc.testing/192.168.130.11:6443
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:270)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:176)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:236)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:109)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:77)
	at okhttp3.internal.connection.Transmitter.newExchange$okhttp(Transmitter.kt:162)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:35)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:82)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:84)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:71)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
	at com.openshift.internal.restclient.okhttp.ResponseCodeInterceptor.intercept(ResponseCodeInterceptor.java:55)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.kt:184)
	at okhttp3.RealCall.execute(RealCall.kt:66)
	at com.openshift.internal.restclient.DefaultClient.request(DefaultClient.java:315)
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:307)
	... 12 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:403)
	at java.base/java.net.Socket.connect(Socket.java:591)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:126)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:268)
	... 35 more
```

**Comment by robstryker** [comment]
> @robstryker @adietish @jeffmaury I would ilke to ask why are we using kubeadmin openshift connection be default... I am thinking that we might be using the other one - developer instead, as it was also used in developers.redhat.com blog about CRC. (https://developers.redhat.com/blog/2019/09/05/red-hat-openshift-4-on-your-laptop-introducing-red-hat-codeready-containers/) WDYT?

When I download a nightly build of CRC, and run it, the output is as follows:

> INFO Then you can access it by running 'oc login -u developer -p developer https://api.crc.testing:6443' 
> INFO To login as an admin, username is 'kubeadmin' and password is *********" 

I guess both would work, and the non-admin connection would probably use much less resources.

**Comment by robstryker** [comment]
> Server Wizard validation: There is substantial improvement, thanks @robstryker. Nevertheless, I still find some issues:
> Passing any executable file (sh script) into "CRC Binary:" labeled text passes the validation, which I think it should not. Should show error saying something like "Please select valid CRC binary/Image" whatever is more precise.
> 
> Server Editor validation is not complete, I can play around with non-existing/directory/non-executable/executable-not-an-crc paths and I still can save the editor without any error/warning dialog appearing.

There is really no way at all for me to know whether a given build or binary is a proper crc binary or not. Seriously. Unless I keep some type of name + filesize + md5 checksum list of approved versions. But even if I did that, then nightly builds wouldn't work and would be rejected. 

I can't make this change. I'm sorry.

**Comment by robstryker** [comment]
> Server Editor validation is not complete, I can play around with non-existing/directory/non-executable/executable-not-an-crc paths and I still can save the editor without any error/warning dialog appearing.

We (or I?) really don't like using the error-dialog when saving bad fields in the server editor unless the fields are likely to cause some type of workspace crash or corruption of the server's home directory ~/.crc or similar.  Otherwise we just mark the fields as an error so the user can see it and fix it. 

Of course the user can still attempt to "start" the server later, with these invalid fields, but we should aim to have the launches or the actions validate those fields before use.

**Comment by robstryker** [comment]
>     5. Insert existing file (pdf)
>        ASSERT: Please select valid image...
>        Actual: we are ok- which is wrong

We will not validate the actual contents of the file. The format may change at any time, the fields may change, and even the type of content may change. Until the stability of these features is shown, we should only do baseline checks for obvious errors.

**Comment by adietish** [comment]
testPR

**Comment by adietish** [comment]
[jenkins build #1212](https://dev-platform-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/jbosstools-openshift-Pull-Request/2112/) failed with test failures in openshift/egit which is a) unrelated b) known to be unstable at times.

**Comment by odockal** [comment]
> > Server Wizard validation: There is substantial improvement, thanks @robstryker. Nevertheless, I still find some issues:
> > Passing any executable file (sh script) into "CRC Binary:" labeled text passes the validation, which I think it should not. Should show error saying something like "Please select valid CRC binary/Image" whatever is more precise.
> > Server Editor validation is not complete, I can play around with non-existing/directory/non-executable/executable-not-an-crc paths and I still can save the editor without any error/warning dialog appearing.
> 
> There is really no way at all for me to know whether a given build or binary is a proper crc binary or not. Seriously. Unless I keep some type of name + filesize + md5 checksum list of approved versions. But even if I did that, then nightly builds wouldn't work and would be rejected.
> 
> I can't make this change. I'm sorry.

I assume that it would be possible at the moment when we try running sub process and checking `crc version` command? The same way we did with minishift?

**Comment by odockal** [comment]
> > @robstryker @adietish @jeffmaury I would ilke to ask why are we using kubeadmin openshift connection be default... I am thinking that we might be using the other one - developer instead, as it was also used in developers.redhat.com blog about CRC. (https://developers.redhat.com/blog/2019/09/05/red-hat-openshift-4-on-your-laptop-introducing-red-hat-codeready-containers/) WDYT?
> 
> When I download a nightly build of CRC, and run it, the output is as follows:
> 
> > INFO Then you can access it by running 'oc login -u developer -p developer https://api.crc.testing:6443'
> > INFO To login as an admin, username is 'kubeadmin' and password is *********"
> 
> I guess both would work, and the non-admin connection would probably use much less resources.

Perfect, thanks!

**Comment by odockal** [comment]
> > Server Editor validation is not complete, I can play around with non-existing/directory/non-executable/executable-not-an-crc paths and I still can save the editor without any error/warning dialog appearing.
> 
> We (or I?) really don't like using the error-dialog when saving bad fields in the server editor unless the fields are likely to cause some type of workspace crash or corruption of the server's home directory ~/.crc or similar. Otherwise we just mark the fields as an error so the user can see it and fix it.
> 
> Of course the user can still attempt to "start" the server later, with these invalid fields, but we should aim to have the launches or the actions validate those fields before use.

I am again replying based on the behavior we used in cdk/minishift where error dialog was thrown once the user tried to save the server editor. 

We also had error marker, what would be the same.

**Comment by odockal** [comment]
> > ```
> > 5. Insert existing file (pdf)
> >    ASSERT: Please select valid image...
> >    Actual: we are ok- which is wrong
> > ```
> 
> We will not validate the actual contents of the file. The format may change at any time, the fields may change, and even the type of content may change. Until the stability of these features is shown, we should only do baseline checks for obvious errors.

That's true, I have to agree to that. thanks for clarification.

**Comment by odockal** [comment]
@robstryker @adietish I would like to continue on jira once we identify proper behavior for crc adapter. Also, we had to keep in mind possible api changes in crc when they release as Rob said... thanks for coop.

**Pull request merged**